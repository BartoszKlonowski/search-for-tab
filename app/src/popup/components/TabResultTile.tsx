import React from "react";
import Browser from "webextension-polyfill";
import {translate} from "../../engine/i18n";

type Props = {
    tabId: number | undefined;
    tabTitle: string;
    onClose: (id: number) => void;
    tabUrl: string;
    favicon?: string;
};

export const TabResultTile = (props: Props) => {
    const {tabId, tabTitle, tabUrl, favicon, onClose} = props;

    const getTabDomainFromURL = (URL: string): string | null => {
        const result = URL.replace("https://", "").replace("http://", "").replace("www.", "");
        const results = result.match(/[^ /]+/g);
        return results && !results[0].includes(" ") && results[0].includes(".") ? results[0] : null;
    };

    const tabDomain = getTabDomainFromURL(tabUrl);

    if (!tabId) {
        return null;
    }

    const makeTabActive = () => {
        Browser.tabs
            .update(tabId, {
                active: true,
            })
            .then(() => {
                window.close();
            })
            .catch((error) => {
                console.error("ERROR: making tab active finished with ", error.message);
            });
    };

    const closeTab = () => {
        Browser.tabs
            .remove(tabId)
            .then(() => {
                onClose(tabId);
            })
            .catch((error) => {
                console.error("ERROR: making tab active finished with ", error.message);
            });
    };

    const trimTitleText = (): string => {
        const maxTitleTextLength = 60;
        return tabTitle.length > maxTitleTextLength
            ? tabTitle.slice(0, maxTitleTextLength).concat("...")
            : tabTitle;
    };

    const getTabIcon = (): string => {
        return favicon ? favicon : `https://icons.duckduckgo.com/ip3/${tabDomain}.ico`;
    };

    return (
        <div className="tab-tile-container">
            <div className="tab-tile-icon-container" title={tabDomain || tabUrl}>
                <img src={getTabIcon()} />
            </div>
            <div
                className="tab-tile-title-container"
                onClick={makeTabActive}
                title={`${translate("go-to")} ${tabDomain}`}>
                <div className="tab-tile-title">{trimTitleText()}</div>
            </div>
            <div
                className="tab-tile-close-action-icon-container"
                onClick={closeTab}
                title={translate("close")}>
                <div>{"+"}</div>
            </div>
        </div>
    );
};
