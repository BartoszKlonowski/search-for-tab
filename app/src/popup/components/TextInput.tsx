import React from "react";

type Props = {
    id: string;
    onChange: () => void;
};

export const TextInput = (props: Props) => {
    return (
        <div className="tab-search-entry-list-container">
            <input
                onChange={props.onChange}
                className="tab-search-entry-list"
                id={props.id}
                placeholder="..."
                type="text"
            />
        </div>
    );
};
