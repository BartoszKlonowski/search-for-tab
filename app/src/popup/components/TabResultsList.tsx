import React, {useMemo, useState} from "react";
import Browser from "webextension-polyfill";
import {TabResultTile} from "./TabResultTile";

type Props = {
    tabSearchPhrase: string;
};

export const TabResultsList = (props: Props) => {
    const {tabSearchPhrase} = props;

    if (!tabSearchPhrase.length) {
        return null;
    }

    const [allTabs, setAllTabs] = useState<Browser.Tabs.Tab[]>();

    const renderSearchResultTile = (
        tabId: number | undefined,
        tabTitle: string,
        tabUrl: string,
        tabFavicon?: string
    ): JSX.Element => {
        return (
            <TabResultTile tabId={tabId} tabTitle={tabTitle} tabUrl={tabUrl} favicon={tabFavicon} />
        );
    };

    useMemo(
        () =>
            Browser.tabs
                .query({currentWindow: true})
                .then((tabs) => {
                    setAllTabs(
                        tabs.filter(
                            (tab) =>
                                tab.title?.includes(tabSearchPhrase) ||
                                tab.url?.includes(tabSearchPhrase)
                        )
                    );
                })
                .catch((error) => {
                    console.error(error);
                }),
        [tabSearchPhrase]
    );

    return (
        <div className="tab-search-results-list-container">
            {allTabs?.map((tab) => (
                <li key={`searchResultTile-${tab.title}`}>
                    {renderSearchResultTile(tab.id, tab.title || "", tab.url || "", tab.favIconUrl)}
                </li>
            ))}
        </div>
    );
};

export default TabResultsList;
