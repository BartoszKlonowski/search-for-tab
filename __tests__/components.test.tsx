import React from "react";
import TestRenderer, {act, ReactTestInstance, ReactTestRenderer} from "react-test-renderer";
import TabResultsList from "../app/src/popup/components/TabResultsList";
import {TabResultTile} from "../app/src/popup/components/TabResultTile";

function renderElement(element: JSX.Element): ReactTestRenderer {
    const component = TestRenderer.create(element);
    return component;
}

async function renderElementAsObject(element: JSX.Element): Promise<ReactTestInstance> {
    let component;
    await act(async () => {
        component = renderElement(element);
    });
    return JSON.parse(JSON.stringify(component.toJSON()));
}

function getChild(renderedObject: ReactTestInstance, childIndex: number): ReactTestInstance {
    const child = renderedObject.children[childIndex] as ReactTestInstance;
    return child;
}

describe("TabResultsList", () => {
    it("renders correctly according to snapshot", async () => {
        global.browser.tabs.query = () => {
            return new Promise((resolve) => {
                resolve([{id: 1, title: "some-results-for-test-phrase"}]);
            });
        };
        const resultsList = await renderElementAsObject(
            <TabResultsList tabSearchPhrase="test-phrase" />
        );
        expect(resultsList).toMatchSnapshot();
    });

    it("renders nothing if empty search phrase is passed within props", async () => {
        const resultsList = await renderElementAsObject(<TabResultsList tabSearchPhrase="" />);
        expect(resultsList).toBeNull();
    });

    it("renders a list when given some search phrase", async () => {
        global.browser.tabs.query = () => {
            return new Promise((resolve) => {
                resolve([{id: 1, title: "some-results-test-search-phrase"}]);
            });
        };
        const resultsList = await renderElementAsObject(
            <TabResultsList tabSearchPhrase="test-search-phrase" />
        );
        expect(resultsList).toBeDefined();
        expect(getChild(resultsList, 0).type).toBe("li");
    });

    it("renders a list of three tabs when exactly three tabs matches the search phrase", async () => {
        global.browser.tabs.query = () => {
            return new Promise((resolve) => {
                resolve([
                    {id: 1, title: "very rarely visited tab"},
                    {id: 2, title: "tab-vary-with-content"},
                    {id: 3, title: "tab-very-interesting"},
                    {id: 4, title: "tab not very empty but matching"},
                    {id: 5, title: "some-result-with-dummy-title"},
                ]);
            });
        };
        const resultsList = await renderElementAsObject(<TabResultsList tabSearchPhrase="very" />);
        expect(resultsList).toBeDefined();
        expect(resultsList.children.length).toBe(3);
    });
});

describe("TabResultTile", () => {
    it("renders correctly according to snapshot", async () => {
        const resultTile = renderElement(
            <TabResultTile
                onClose={(id: number) => {
                    id;
                }}
                tabId={1}
                tabTitle="test-phrase"
                tabUrl="test-url"
            />
        );
        expect(resultTile.toJSON()).toMatchSnapshot();
    });

    it("uses the correct domain for a certain tab URL", async () => {
        const resultTileWithDefaultURL = await renderElementAsObject(
            <TabResultTile
                onClose={(id: number) => {
                    id;
                }}
                tabId={1}
                tabTitle="fake-title"
                tabUrl="www.google.com"
            />
        );
        let resultTileIcon = getChild(getChild(resultTileWithDefaultURL, 0), 0);
        expect(resultTileIcon.props.src).toBe("https://icons.duckduckgo.com/ip3/google.com.ico");
        const resultTileWithHTTPSURL = await renderElementAsObject(
            <TabResultTile
                onClose={(id: number) => {
                    id;
                }}
                tabId={1}
                tabTitle="fake-title"
                tabUrl="https://google.com"
            />
        );
        resultTileIcon = getChild(getChild(resultTileWithHTTPSURL, 0), 0);
        expect(resultTileIcon.props.src).toBe("https://icons.duckduckgo.com/ip3/google.com.ico");
        const resultTileWithComplexURL = await renderElementAsObject(
            <TabResultTile
                onClose={(id: number) => {
                    id;
                }}
                tabId={1}
                tabTitle="fake-title"
                tabUrl="http://google.com/very-complex?url"
            />
        );
        resultTileIcon = getChild(getChild(resultTileWithComplexURL, 0), 0);
        expect(resultTileIcon.props.src).toBe("https://icons.duckduckgo.com/ip3/google.com.ico");
    });

    it("uses the default favicon if exists", async () => {
        const resultTileWithDefaultURL = await renderElementAsObject(
            <TabResultTile
                onClose={(id: number) => {
                    id;
                }}
                tabId={1}
                tabTitle="fake-title"
                favicon="existing-correct-icon.ico"
                tabUrl="www.google.com"
            />
        );
        const resultTileIcon = getChild(getChild(resultTileWithDefaultURL, 0), 0);
        expect(resultTileIcon.props.src).toBe("existing-correct-icon.ico");
    });

    it("does not render when the ID of the tab is broken", async () => {
        let resultTile = await renderElementAsObject(
            <TabResultTile
                onClose={(id: number) => {
                    id;
                }}
                tabId={0}
                tabTitle="fake-title"
                tabUrl="fake-tab-url"
            />
        );
        expect(resultTile).toBeNull();
        resultTile = await renderElementAsObject(
            <TabResultTile
                onClose={(id: number) => {
                    id;
                }}
                tabId={undefined}
                tabTitle="fake-title"
                tabUrl="fake-tab-url"
            />
        );
        expect(resultTile).toBeNull();
    });

    it("goes to selected tab by making it active", async () => {
        global.browser.tabs.update = () => {
            return new Promise(() => {
                return true;
            });
        };
        const tabTile = await renderElementAsObject(
            <TabResultTile
                onClose={(id: number) => {
                    id;
                }}
                tabId={11345}
                tabTitle="fake-title"
                tabUrl="fake-tab-url"
            />
        );
        const tabTileClickableContainer = getChild(tabTile, 1);
        expect(tabTileClickableContainer.props).toBeDefined();
    });

    it("closes selected tab by clicking close icon", async () => {
        global.browser.tabs.update = () => {
            return new Promise(() => {
                return true;
            });
        };
        const tabTile = await renderElementAsObject(
            <TabResultTile
                onClose={(id: number) => {
                    id;
                }}
                tabId={11345}
                tabTitle="fake-title"
                tabUrl="fake-tab-url"
            />
        );
        const tabTileClickableContainer = getChild(tabTile, 2);
        expect(tabTileClickableContainer.props).toBeDefined();
    });
});
