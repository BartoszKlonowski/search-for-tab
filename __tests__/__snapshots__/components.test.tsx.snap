// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabResultTile renders correctly according to snapshot 1`] = `
<div
  className="tab-tile-container"
>
  <div
    className="tab-tile-icon-container"
    title="test-url"
  >
    <img
      src="https://icons.duckduckgo.com/ip3/null.ico"
    />
  </div>
  <div
    className="tab-tile-title-container"
    onClick={[Function]}
    title="Go to: null"
  >
    <div
      className="tab-tile-title"
    >
      test-phrase
    </div>
  </div>
  <div
    className="tab-tile-close-action-icon-container"
    onClick={[Function]}
    title="Close"
  >
    <div>
      +
    </div>
  </div>
</div>
`;

exports[`TabResultsList renders correctly according to snapshot 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": null,
                  "props": Object {
                    "src": "https://icons.duckduckgo.com/ip3/null.ico",
                  },
                  "type": "img",
                },
              ],
              "props": Object {
                "className": "tab-tile-icon-container",
                "title": "",
              },
              "type": "div",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    "some-results-for-test-phrase",
                  ],
                  "props": Object {
                    "className": "tab-tile-title",
                  },
                  "type": "div",
                },
              ],
              "props": Object {
                "className": "tab-tile-title-container",
                "title": "Go to: null",
              },
              "type": "div",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    "+",
                  ],
                  "props": Object {},
                  "type": "div",
                },
              ],
              "props": Object {
                "className": "tab-tile-close-action-icon-container",
                "title": "Close",
              },
              "type": "div",
            },
          ],
          "props": Object {
            "className": "tab-tile-container",
          },
          "type": "div",
        },
      ],
      "props": Object {},
      "type": "li",
    },
  ],
  "props": Object {
    "className": "tab-search-results-list-container",
  },
  "type": "div",
}
`;
